src/cli/pod.toit

```diff
@@ -213,3 +228,74 @@ class Pod:
       parts[part-name] = file.content
     manifest["parts"] = part-names
     block.call manifest parts
+
+  /**
+  Computes the device-specific data of the given envelope.
+
+  Combines this pod and identity ($identity-path) into a single firmware image
+    and computes the configuration which depends on the checksums of the
+    individual parts.
+
+  In this context the configuration consists of the checksums of the individual
+    parts of the firmware image, combined with the configuration that was
+    stored in the envelope.
+  */
+  compute-device-specific-data -> ByteArray
+      --identity-path/string
+      --cache/Cache
+      --ui/Ui:
+    // Use the SDK from the pod.
+    sdk := get-sdk sdk-version --cache=cache
+
+    // Extract the device ID from the identity file.
+    // TODO(florian): abstract the identity management.
+    identity-raw := file.read-content identity-path
+
+    identity := ubjson.decode (base64.decode identity-raw)
+
+    // Since we already have the identity content, check that the artemis server
+    // is the same.
+    // This is primarily a sanity check, and we might remove the broker from the
+    // identity file in the future. Since users are not supposed to be able to
+    // change the Artemis server, there wouldn't be much left of the check.
+    // TODO(florian): remove this check?
+    with-tmp-directory: | tmp/string |
+      artemis-assets-path := "$tmp/artemis.assets"
+      sdk.run-firmware-tool [
+        "-e", envelope-path,
+        "container", "extract",
+        "-o", artemis-assets-path,
+        "--part", "assets",
+        "artemis"
+      ]
+
+      if not is-same-broker "broker" identity tmp artemis-assets-path sdk:
+        ui.warning "The identity file and the Artemis assets in the envelope don't use the same broker"
+      if not is-same-broker "artemis.broker" identity tmp artemis-assets-path sdk:
+        ui.warning "The identity file and the Artemis assets in the envelope don't use the same Artemis server"
+
+    device-map := identity["artemis.device"]
+    device := Device
+        --hardware-id=uuid.parse device-map["hardware_id"]
+        --id=uuid.parse device-map["device_id"]
+        --organization-id=uuid.parse device-map["organization_id"]
+
+    // We don't really need the full firmware and just the device-specific data,
+    // but by cooking the firmware we get the checksums correct.
+    firmware := Firmware --pod=this --device=device --cache=cache
```

Comment:
This `Firmware` instance seems useful for uploading the device-specific trivial patches. 

Maybe we should really return the device-specific firmware from this in the next PR?

Reply:
I put it onto my TODO list.